src/server/services/procedures/thing-speak/thing-speak.js

```diff
@@ -49,73 +47,105 @@ function detailParser(item) {
     const first = await this._requestData({ path, queryString: rpcUtils.encodeQueryData(query) });
     const totalPages = Math.ceil(first.pagination.total_entries / first.pagination.per_page);
-    const pages = Math.min(totalPages, Math.ceil(limit / first.pagination.per_page));
-    
-    const requests = [];
-    for (query.page = 1; query.page <= pages; ++query.page) {
-        requests.push(this._requestData({ path, queryString: rpcUtils.encodeQueryData(query) }));
-    }
-    const results = await Promise.all(requests);
+    const pagesPerIter = 8;
 
-    const items = [];
-    for (const res of results) {
-        for (const item of res.channels) {
-            items.push(detailParser(item));
-            if (items.length >= limit) break;
+    const final = [];
+    for (let iter = 1; iter <= 128 && query.page <= totalPages; ++iter) {
+        const endPage = Math.min(query.page + pagesPerIter, totalPages);
+        const requests = [];
+        for (; query.page <= endPage; ++query.page) {
+            requests.push(this._requestData({ path, queryString: rpcUtils.encodeQueryData(query) }));
         }
+        const results = await Promise.all(requests);
+
+        const items = [];
+        for (const res of results) {
+            for (const item of res.channels) {
+                items.push(item);
+            }
+        }
+
+        const beforeInclude = final.length;
+        const keepFlags = await Promise.all(items.map(pred));
+        for (let i = 0; i < items.length; ++i) {
+            if (!keepFlags[i]) continue;
+            final.push(detailParser(items[i]));
+            if (final.length >= limit) return final;
+        }
+
+        const keptFrac = (final.length - beforeInclude) / items.length;
+        this._logger.info(`paginatedSearch - ${final.length} items after iter ${iter} (kept ${100 * keptFrac}%)`);
     }
-    return items;
+    return final;
 };
 
 /**
  * Search for ThingSpeak channels by tag.
  *
- * @param {String} tag
- * @param {Number=} limit
+ * @param {String} tag tag to search for
+ * @param {Number=} limit max number of results to return
+ * @param {Date=} updatedSince only include results which have (some) new data since this date
+ * @returns {Array<Object>} search results
  */
-thingspeakIoT.searchByTag = function(tag, limit) {
-    return this._paginatedSearch('public.json', { tag: encodeURIComponent(tag) }, limit);
+thingspeakIoT.searchByTag = function(tag, limit=15, updatedSince=null) {
+    const pred = getUpdatedSincePred(updatedSince);
+    return this._paginatedSearch('public.json', { tag: encodeURIComponent(tag) }, limit, pred);
 };
 
 /**
  * Search for channels by location.
  *
- * @param {Latitude} latitude
- * @param {Longitude} longitude
- * @param {BoundedNumber<0>=} distance
- * @param {Number=} limit
+ * @param {Latitude} latitude latitude to search near
+ * @param {Longitude} longitude longitude to search near
+ * @param {BoundedNumber<0>=} distance max distance from location.
```

Comment:
@dragazo Would it be possible to add the defaults to the documentation strings (unless distance/limit are at their maximums)?

Reply:
@gsteinLTU Sure, new commit adds that.